llvm-external-projects/iree-dialects/test/Dialect/iree_linalg_ext/tiling.mlir

```diff
@@ -916,3 +916,190 @@ func.func @topk_tile_tensor_optional(%input_values: tensor<20x10xf32>, %out_valu
+// CHECK:               %[[SUB_RES:.*]] = iree_linalg_ext.pack
+// CHECK-SAME:            {__internal_linalg_transform__ = "tiling_pack_output"}
+// CHECK-SAME:            %[[SUB_IN]] dims_pos = [0, 1] inner_tiles = [32, 32] into %[[SUB_OUT]]
+// CHECK:               %[[INSERT:.*]] = tensor.insert_slice %[[SUB_RES]] into %[[ITER1]]
+// CHECK:               scf.yield %[[INSERT]] : tensor<4x8x32x32xf32>
+// CHECK:             }
+// CHECK:             scf.yield %[[RES1:.*]] : tensor<4x8x32x32xf32>
+// CHECK:           }
+// CHECK:           return %[[RES0:.*]] : tensor<4x8x32x32xf32>
+// CHECK:         }
+
+// -----
+
+func.func @pad_and_pack_static(%input: tensor<13x15xf32>, %output: tensor<2x8x8x2xf32>, %pad: f32) -> tensor<2x8x8x2xf32> {
+  %0 = iree_linalg_ext.pack {__internal_linalg_transform__ = "tiling_pack_input"} %input padding_value(%pad : f32) dims_pos = [0, 1] inner_tiles = [8, 2] into %output : (tensor<13x15xf32> tensor<2x8x8x2xf32>) -> tensor<2x8x8x2xf32>
+  return %0 : tensor<2x8x8x2xf32>
+}
+// CHECK-DAG:     #[[MAP0:.+]] = affine_map<(d0)[s0, s1] -> (2, -d0 + s1)>
+// CHECK-DAG:     #[[MAP1:.+]] = affine_map<(d0)[s0, s1] -> (4, -d0 + s1)>
+// CHECK-DAG:     #[[MAP2:.+]] = affine_map<(d0) -> (d0 * 8)>
+// CHECK-DAG:     #[[MAP3:.+]] = affine_map<(d0, d1) -> (d0 - d1)>
+// CHECK-DAG:     #[[MAP4:.+]] = affine_map<(d0, d1) -> (d0, d1)>
+// CHECK-DAG:     #[[MAP5:.+]] = affine_map<(d0) -> (d0 * 2)>
+// CHECK-LABEL:   func.func @pad_and_pack_static(
+// CHECK-SAME:                                   %[[IN:.*]]: tensor<13x15xf32>,
+// CHECK-SAME:                                   %[[OUT:.*]]: tensor<2x8x8x2xf32>,
+// CHECK-SAME:                                   %[[PAD:.*]]: f32) -> tensor<2x8x8x2xf32> {
+// CHECK-DAG:       %[[C0:.*]] = arith.constant 0 : index
+// CHECK-DAG:       %[[C2:.*]] = arith.constant 2 : index
+// CHECK-DAG:       %[[C4:.*]] = arith.constant 4 : index
+// CHECK-DAG:       %[[C8:.*]] = arith.constant 8 : index
+// CHECK-DAG:       %[[C13:.*]] = arith.constant 13 : index
+// CHECK-DAG:       %[[C15:.*]] = arith.constant 15 : index
+// CHECK:           %[[RES0:.*]] = scf.for %[[I:.*]] = %[[C0]] to %[[C2]] step %[[C2]] iter_args(%[[ITER0:.*]] = %[[OUT]]) -> (tensor<2x8x8x2xf32>) {
+// CHECK-DAG:         %[[OUT_I_SZ:.*]] = affine.min #[[MAP0]](%[[I]])
+// CHECK:             %[[RES1:.*]] = scf.for %[[J:.*]] = %[[C0]] to %[[C8]] step %[[C4]] iter_args(%[[ITER1:.*]] = %[[ITER0]]) -> (tensor<2x8x8x2xf32>) {
+// CHECK-DAG:           %[[OUT_J_SZ:.*]] = affine.min #[[MAP1]](%[[J]])
+// CHECK-DAG:           %[[IN_I:.*]] = affine.apply #[[MAP2]](%[[I]])
+// CHECK-DAG:           %[[IN_I_SZ:.*]] = affine.min #[[MAP4]]
+// CHECK-DAG:           %[[IN_J:.*]] = affine.apply #[[MAP5]](%[[J]])
+// CHECK-DAG:           %[[IN_J_SZ:.*]] = affine.min #[[MAP4]]
+// CHECK:               %[[SUB_IN:.*]] = tensor.extract_slice %[[IN]][%[[IN_I]], %[[IN_J]]] [%[[IN_I_SZ]], %[[IN_J_SZ]]] [1, 1] : tensor<13x15xf32> to tensor<?x?xf32>
```

Comment:
I'd expect an `scf.if` for handling boundary cases? (the `tensor.pad` op tiling implementation does this). Basically you dont want to end up having the size evaluate to zero in the `tensor.extract_slice`. Thats UB. So you need an `if` condition for when the result of the tile is just pure padding, in which case you generate a tensor using `tensor.generate` (or whatever the `tensor.pad` does.

Reply:
The tensor.generate is needed when the original data source is not read at all. E.g., The tile locate at bottom right corner and the high values are huge. In this case, we load nothing from source and use tensor.generate to create a tensor. It is not the case in linalg_ext.pack op because the in-complete tile must contains some of inputs.

In tiling pad op case, we also use apply_min to handle boundary cases. The logic is more complicated because they have low padding. E.g.,

```mlir
#map0 = affine_map<()[s0] -> (s0 + 8)>
#map1 = affine_map<()[s0] -> (s0 + 7)>
#map2 = affine_map<(d0)[s0, s1] -> (2, -d0 + s1)>
#map3 = affine_map<(d0)[s0, s1] -> (3, -d0 + s1)>
#map4 = affine_map<(d0, d1) -> (d0 - d1)>
#map5 = affine_map<(d0, d1) -> (0, d1)>
#map6 = affine_map<(d0, d1) -> (d0, 0)>
#map7 = affine_map<(d0, d1) -> (d0, d1)>
#map8 = affine_map<(d0, d1) -> (d0 + d1)>
...
        %8 = affine.min #map3(%arg4)[%c3, %3]
        %9 = affine.apply #map4(%c3, %arg2)
        %10 = affine.max #map5(%c0, %9)
        %11 = affine.apply #map4(%arg2, %c3)
        %12 = affine.max #map6(%11, %c0)
        %13 = affine.min #map7(%12, %0)
        %14 = affine.apply #map8(%11, %6)
        %15 = affine.max #map6(%14, %c0)
        %16 = affine.min #map7(%15, %0)
        %17 = affine.apply #map4(%16, %13)
        %29 = affine.apply #map4(%28, %25)
        %34 = scf.if %31 -> (tensor<?x?xf32>) {
          %36 = tensor.generate %6, %8 {
          ^bb0(%arg6: index, %arg7: index):
            tensor.yield %arg1 : f32
          } : tensor<?x?xf32>
          scf.yield %36 : tensor<?x?xf32>
        } else {
          %36 = tensor.extract_slice %arg0[%13, %25] [%17, %29] [1, 1] : tensor<?x?xf32> to tensor<?x?xf32>
          %37 = tensor.pad %36 low[%10, %22] high[%20, %33] {
          ^bb0(%arg6: index, %arg7: index):
            tensor.yield %arg1 : f32
          } : tensor<?x?xf32> to tensor<?x?xf32>
          scf.yield %37 : tensor<?x?xf32>
        } {__internal_linalg_transform__ = "tiled"}
```

Reply:
very interesting. I see your point now. We can revisit if the need arises.

Reply:
I just realized that we could need an `scf.if` for handling boundary cases. It could happen when tiling sizes are less than inner_tiles... I'll think more about it and maybe will send out some fixes.